blog/_posts/2018-11-23-ml-compiler.md

```diff
@@ -0,0 +1,69 @@
+---
+layout: post
+title:  Building a Language and Compiler for Machine Learning
+author: Mike Innes, Elliot Saba, Keno Fischer, Dhairya Gandhi, Marco Concetto Rudilosso, Neethu Mariya Joy, Tejan Karmali, Avik Pal, Matt Kelley, Viral Shah
+---
+
+# Building a Language and Compiler for Machine Learning
+
+Since we [originally proposed](https://julialang.org/blog/2017/12/ml&pl) the need for a first-class language, compiler and ecosystem for machine learning (ML), there have been plenty of interesting developments in the field. Not only have the tradeoffs in existing systems, such as TensorFlow and PyTorch, not been resolved, but they are clearer than ever now that both frameworks contain distinct [“static graph”](https://pytorch.org/docs/master/jit.html) and [“eager execution”](https://www.tensorflow.org/guide/eager) interfaces. Meanwhile, the idea of ML models fundamentally being differentiable algorithms – often called [differentiable programming](https://www.facebook.com/yann.lecun/posts/10155003011462143) – has caught on.
+
+Where current frameworks fall short, several exciting new projects have sprung up that dispense with graphs entirely, to bring differentiable programming to the mainstream. [Myia](https://github.com/mila-udem/myia), by the Theano team, differentiates and compiles a subset of Python to high-performance GPU code. [Swift for TensorFlow](https://github.com/tensorflow/swift) extends Swift so that compatible functions can be compiled to TensorFlow graphs. And finally, the [Flux](https://github.com/FluxML/Flux.jl) ecosystem is extending Julia’s compiler with a number of ML-focused tools, including first-class gradients, just-in-time CUDA kernel compilation, automatic batching and support for new hardware such as TPUs.
+
+All of these projects have enormous potential, but we think Julia has an edge. This post, based on our [paper to be presented at NIPS MLSys](https://arxiv.org/abs/1811.01457), will explore how we have used Julia to re-think ML tooling from the ground up, and provides some insight into the work that modern ML frameworks need to do.
+
+## Enter Flux
+
+We need a language to write differentiable algorithms, and Flux takes Julia to be this language. Being designed from the ground up for mathematical and numerical computing, Julia is unusually well-suited for expressing ML algorithms.  Meanwhile, its mix of modern design and new ideas in the compiler makes it easier to address the high performance needs of cutting edge ML.
+
+Where typical frameworks are all-encompassing monoliths in millions of lines of C++, Flux is only a thousand lines of straightforward Julia code that plays well with the rest of the ecosystem. Simply take [one package for taking gradients](https://github.com/FluxML/Zygote.jl), [one package for GPU support](https://github.com/JuliaGPU/CuArrays.jl/), sprinkle with some light convenience functions, bake for fifteen minutes and out pops a fully-featured ML stack.
```

Comment:
TensorFlow is 1.5 million lines of C++; PyTorch is about 500,000 (including 220k in the Caffe2 tree). Still absolutely insane but not at the level of e.g. LLVM (8 million); and probably best to say "hundreds of thousands" rather than millions?

Reply:
Ah, the old prosody/accuracy tradeoff. I've updated the post.

Reply:
Could do “zillions” :)